test/common/ssl/context_impl_test.cc

```diff
@@ -155,16 +210,66 @@ TEST_F(SslContextImplTest, TestGetCertInformation) {
-  std::string cert_chain_partial_output(
-      TestEnvironment::substitute("Certificate Path: {{ test_tmpdir }}/unittestcert.pem"));
+  std::string ca_cert_json = R"EOF({
+ "path": "{{ test_rundir }}/test/common/ssl/test_data/ca_cert.pem",
+ "serial_number": "eaf3b0ea1d0e579a",
+ "subject_alt_names": [],
+)EOF";
+
+  std::string cert_chain_json = R"EOF({
+ "path": "{{ test_tmpdir }}/unittestcert.pem",
+)EOF";
+
+  std::string ca_cert_partial_output(TestEnvironment::substitute(ca_cert_json));
+  std::string cert_chain_partial_output(TestEnvironment::substitute(cert_chain_json));
+  EXPECT_TRUE(MessageUtil::getJsonStringFromMessage(*context->getCaCertInformation(), true, true)
+                  .find(ca_cert_partial_output) != std::string::npos);
+  EXPECT_TRUE(MessageUtil::getJsonStringFromMessage(*context->getCertChainInformation(), true, true)
+                  .find(cert_chain_partial_output) != std::string::npos);
+}
+
+TEST_F(SslContextImplTest, TestGetCertInformationWithSAN) {
+  std::string json = R"EOF(
+  {
+    "cert_chain_file": "{{ test_rundir }}/test/common/ssl/test_data/san_dns_chain3.pem",
+    "private_key_file": "{{ test_rundir }}/test/common/ssl/test_data/san_dns_key3.pem",
+    "ca_cert_file": "{{ test_rundir }}/test/common/ssl/test_data/san_dns_cert3.pem"
+  }
+  )EOF";
 
-  EXPECT_TRUE(context->getCaCertInformation().find(ca_cert_partial_output) != std::string::npos);
-  EXPECT_TRUE(context->getCertChainInformation().find(cert_chain_partial_output) !=
-              std::string::npos);
+  Json::ObjectSharedPtr loader = TestEnvironment::jsonLoadFromString(json);
+  ClientContextConfigImpl cfg(*loader, factory_context_);
+  Runtime::MockLoader runtime;
+  ContextManagerImpl manager(runtime);
+  Stats::IsolatedStoreImpl store;
+
+  ClientContextSharedPtr context(manager.createSslClientContext(store, cfg));
+  std::string ca_cert_json = R"EOF({
+ "path": "{{ test_rundir }}/test/common/ssl/test_data/san_dns_cert3.pem",
+ "serial_number": "b13ff63f2dbc118d",
+ "subject_alt_names": [
+  {
+   "dns": "server1.example.com"
+  }
+ ],
+)EOF";
+
+  std::string cert_chain_json = R"EOF({
+ "path": "{{ test_rundir }}/test/common/ssl/test_data/san_dns_chain3.pem",
+)EOF";
+
+  // This is similar to the hack above, but right now we generate the ca_cert and it expires in 15
+  // days only in the first second that it's valid. We will partially match for up until Days until
+  // Expiration: 1.
+  // For the cert_chain, it is dynamically created when we run_envoy_test.sh which changes the
+  // serial number with
+  // every build. For cert_chain output, we check only for the certificate path.
+  std::string ca_cert_partial_output(TestEnvironment::substitute(ca_cert_json));
+  std::string cert_chain_partial_output(TestEnvironment::substitute(cert_chain_json));
+  EXPECT_TRUE(MessageUtil::getJsonStringFromMessage(*context->getCaCertInformation(), true, true)
+                  .find(ca_cert_partial_output) != std::string::npos);
```

Comment:
Use MessageDifferencer instead of comparing JSON string.

Reply:
I tried to use MessageDifferencer and setup is becoming complicated because we need to ignore certain fields like days_until_expired etc as they change. So I left this with string comparison.

Reply:
You should be able to do that with PARTIAL scope. https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.util.message_differencer#MessageDifferencer.Scope.details